src/main/java/gift/repository/ProductRepositoryImpl.java

```diff
@@ -3,75 +3,120 @@
+import org.springframework.jdbc.core.simple.JdbcClient;
+import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
 import org.springframework.stereotype.Repository;
 import org.springframework.web.server.ResponseStatusException;
 
+import javax.sql.DataSource;
 import java.util.*;
 
 @Repository
 public class ProductRepositoryImpl implements ProductRepository{
 
-    private final Map<Long, Product> products = new HashMap<>();
-    private Long listId;
+    private final JdbcClient jdbcClient;
+    private final SimpleJdbcInsert jdbcInsert;
 
-    public ProductRepositoryImpl() {
-        this.listId = 0L;
+    public ProductRepositoryImpl(DataSource dataSource) {
+        this.jdbcClient = JdbcClient.create(dataSource);
+        this.jdbcInsert = new SimpleJdbcInsert(dataSource)
+                .withTableName("products")
+                .usingGeneratedKeyColumns("id");
     }
 
     @Override
     public List<ProductResponseDto> findAllProducts() {
 
-        return products.values()
-                .stream()
-                .map(product -> new ProductResponseDto(
-                        product.getId(),
-                        product.getName(),
-                        product.getPrice(),
-                        product.getImageUrl()
-                ))
-                .toList();
+        String sql = "select * from products";
+        List<ProductResponseDto> result = jdbcClient.sql(sql)
+                .query(ProductResponseDto.class)
+                .list();
+
+        return result;
     }
 
     @Override
-    public Product findProductById (Long id) {
+    public Optional<Product> findProductById (Long id) {
+
+        String sql = "select * from products where id = ?";
+        Optional<Product> result = jdbcClient.sql(sql)
+                .param(id)
+                .query(Product.class)
+                .optional();
 
-        return products.get(id);
+        return result;
     }
 
     @Override
     public Product findProductByIdElseThrow(Long id) {
 
-        Product product = products.get(id);
-
-        if (product == null)
-            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "해당 ID의 상품은 존재하지 않습니다.");
-
-        return product;
+        String sql = "select * from products where id = ?";
+        Product result = jdbcClient.sql(sql)
+                .param(id)
+                .query(Product.class)
+                .optional()
+                .orElseThrow(() ->
+                        new ResponseStatusException(
+                                HttpStatus.NOT_FOUND,
+                                "해당 ID의 상품은 존재하지 않습니다."
+                        )
+                );
+
+        return result;
     }
 
     @Override
     public Product saveProduct(String name, Long price, String imageUrl) {
 
-        listId++;
+        Map<String, Object> params = new HashMap<>();
+
+        params.put("name", name);
+        params.put("price", price);
+        params.put("imageUrl", imageUrl);
```

Comment:
```suggestion
        final Map<String, Object> params = Map.of(
                "name", name,
                "price", price,
                "imageUrl", imageUrl
        );
```
지금 코드도 좋지만 개인적으로 이렇게 불변객체로 선언하는 방식을 선호합니다.
어떤 장점이 있을까요~?